File: src/robotocore/services/lambda_/runtimes/bootstraps/bootstrap.js
/**
 * Node.js Lambda bootstrap — reads event from stdin, invokes handler, writes result to stdout.
 *
 * Usage: node bootstrap.js <handler>
 * Where handler is "module.function" (e.g., "index.handler")
 *
 * Protocol:
 *   stdin  -> JSON event
 *   stdout -> JSON result
 *   stderr -> logs
 */

const path = require('path');

async function main() {
    const handlerSpec = process.env._HANDLER || process.argv[2];
    if (!handlerSpec) {
        const err = { errorMessage: 'No handler specified', errorType: 'Runtime.HandlerNotFound' };
        process.stdout.write(JSON.stringify(err));
        process.exit(1);
    }

    const dotIndex = handlerSpec.lastIndexOf('.');
    if (dotIndex < 0) {
        const err = { errorMessage: `Bad handler format: ${handlerSpec}`, errorType: 'Runtime.HandlerNotFound' };
        process.stdout.write(JSON.stringify(err));
        process.exit(1);
    }

    const modulePath = handlerSpec.substring(0, dotIndex);
    const funcName = handlerSpec.substring(dotIndex + 1);

    // Read event from stdin
    let inputData = '';
    for await (const chunk of process.stdin) {
        inputData += chunk;
    }

    let event;
    try {
        event = JSON.parse(inputData);
    } catch (e) {
        event = {};
    }

    // Build context
    const context = {
        functionName: process.env.AWS_LAMBDA_FUNCTION_NAME || 'test-function',
        functionVersion: '$LATEST',
        invokedFunctionArn: `arn:aws:lambda:${process.env.AWS_REGION || 'us-east-1'}:${process.env.AWS_ACCOUNT_ID || '123456789012'}:function:${process.env.AWS_LAMBDA_FUNCTION_NAME || 'test-function'}`,
        memoryLimitInMB: process.env.AWS_LAMBDA_FUNCTION_MEMORY_SIZE || '128',
        awsRequestId: require('crypto').randomUUID(),
        logGroupName: `/aws/lambda/${process.env.AWS_LAMBDA_FUNCTION_NAME || 'test-function'}`,
        logStreamName: `${new Date().toISOString().split('T')[0].replace(/-/g, '/')}/$LATEST`,
        getRemainingTimeInMillis: () => {
            const timeout = parseInt(process.env.AWS_LAMBDA_FUNCTION_TIMEOUT || '3', 10);
            return timeout * 1000;
        },
        done: (err, result) => { if (err) throw err; return result; },
        succeed: (result) => result,
        fail: (err) => { throw err; },
    };

    // Load the handler module
    let handlerModule;
    try {
        const fullPath = path.resolve(process.cwd(), modulePath);
        handlerModule = require(fullPath);
    } catch (e) {
        const err = {
            errorMessage: `Cannot find module '${modulePath}': ${e.message}`,
            errorType: 'Runtime.ImportModuleError',
        };
        process.stdout.write(JSON.stringify(err));
        process.exit(1);
    }

    const handlerFunc = handlerModule[funcName];
    if (typeof handlerFunc !== 'function') {
        const err = {
            errorMessage: `Handler '${funcName}' is not a function in '${modulePath}'`,
            errorType: 'Runtime.HandlerNotFound',
        };
        process.stdout.write(JSON.stringify(err));
        process.exit(1);
    }

    // Invoke the handler
    try {
        let result;
        if (handlerFunc.length <= 2) {
            // async handler or 2-arg handler
            result = await handlerFunc(event, context);
        } else {
            // callback-style handler(event, context, callback)
            result = await new Promise((resolve, reject) => {
                handlerFunc(event, context, (err, res) => {
                    if (err) reject(err);
                    else resolve(res);
                });
            });
        }
        process.stdout.write(JSON.stringify(result === undefined ? null : result));
    } catch (e) {
        const err = {
            errorMessage: e.message || String(e),
            errorType: e.constructor ? e.constructor.name : 'Error',
            stackTrace: (e.stack || '').split('\n'),
        };
        process.stdout.write(JSON.stringify(err));
        process.exit(1);
    }
}

main().catch(e => {
    process.stderr.write(`Bootstrap error: ${e.message}\n`);
    process.exit(1);
});

File: src/robotocore/services/lambda_/runtimes/bootstraps/bootstrap.rb
# Ruby Lambda bootstrap — reads event from stdin, invokes handler, writes result to stdout.
#
# Usage: ruby bootstrap.rb
# Handler is read from _HANDLER env var (e.g., "lambda_function.handler")
#
# Protocol:
#   stdin  -> JSON event
#   stdout -> JSON result
#   stderr -> logs

require 'json'
require 'securerandom'

handler_spec = ENV['_HANDLER'] || ARGV[0]
unless handler_spec
  $stdout.write(JSON.generate({
    errorMessage: 'No handler specified',
    errorType: 'Runtime.HandlerNotFound'
  }))
  exit(1)
end

dot_index = handler_spec.rindex('.')
unless dot_index
  $stdout.write(JSON.generate({
    errorMessage: "Bad handler format: #{handler_spec}",
    errorType: 'Runtime.HandlerNotFound'
  }))
  exit(1)
end

module_name = handler_spec[0...dot_index]
method_name = handler_spec[dot_index + 1..]

# Read event from stdin
input_data = $stdin.read
event = begin
  JSON.parse(input_data)
rescue JSON::ParserError
  {}
end

# Build context
context_data = {
  function_name: ENV['AWS_LAMBDA_FUNCTION_NAME'] || 'test-function',
  function_version: '$LATEST',
  invoked_function_arn: "arn:aws:lambda:#{ENV['AWS_REGION'] || 'us-east-1'}:#{ENV['AWS_ACCOUNT_ID'] || '123456789012'}:function:#{ENV['AWS_LAMBDA_FUNCTION_NAME'] || 'test-function'}",
  memory_limit_in_mb: (ENV['AWS_LAMBDA_FUNCTION_MEMORY_SIZE'] || '128').to_i,
  aws_request_id: SecureRandom.uuid,
  log_group_name: "/aws/lambda/#{ENV['AWS_LAMBDA_FUNCTION_NAME'] || 'test-function'}",
  log_stream_name: "#{Time.now.strftime('%Y/%m/%d')}/$LATEST",
}

# Simple context object with method access
LambdaContext = Struct.new(*context_data.keys, keyword_init: true) do
  def get_remaining_time_in_millis
    timeout = (ENV['AWS_LAMBDA_FUNCTION_TIMEOUT'] || '3').to_i
    timeout * 1000
  end
end
context = LambdaContext.new(**context_data)

# Load the handler module
module_file = File.join(Dir.pwd, "#{module_name}.rb")
unless File.exist?(module_file)
  $stdout.write(JSON.generate({
    errorMessage: "Cannot find module '#{module_name}': #{module_file}",
    errorType: 'Runtime.ImportModuleError'
  }))
  exit(1)
end

begin
  require module_file
rescue => e
  $stdout.write(JSON.generate({
    errorMessage: "Error loading module '#{module_name}': #{e.message}",
    errorType: 'Runtime.ImportModuleError',
    stackTrace: e.backtrace || []
  }))
  exit(1)
end

# Find the handler method (top-level defs are private in Ruby, use respond_to? with true)
unless respond_to?(method_name.to_sym, true)
  $stdout.write(JSON.generate({
    errorMessage: "Handler method '#{method_name}' not found in '#{module_name}'",
    errorType: 'Runtime.HandlerNotFound'
  }))
  exit(1)
end

# Invoke
begin
  result = send(method_name.to_sym, event: event, context: context)
  $stdout.write(JSON.generate(result))
rescue => e
  $stderr.write("#{e.class}: #{e.message}\n#{e.backtrace&.join("\n")}\n")
  $stdout.write(JSON.generate({
    errorMessage: e.message,
    errorType: e.class.name,
    stackTrace: e.backtrace || []
  }))
  exit(1)
end
